Bug 1947806: test/extended/router: enable http/2, gRPC router tests

frobware opened the pull request:
Needs openshift/cluster-ingress-operator#585.

This PR re-enables the http/2 & gRPC tests. They were previously 
disabled as the default ingresscontroller is not configured with http/2
support by default.

There are a number of changes here:

- The text fixture has been split up so that we can first apply the
pod/service, then apply the configured routes, then roll out a
dedicated ingresscontroller.

- There is a new fixture to rollout a new router shard with http/2
enabled.

- There is a new test fixture for the routes which are now parameterised
with an explicit domain and namespace selector to match on the router
shard type.

- I've deleted all the old test-fixture generation.

- The test now spins up a dedicated ingresscontroller.

frobware commented:
/test e2e-ovirt

does this work?

frobware commented:
googleapi: Error 403: Quota exceeded for quota group 'ListGroup' 

/retest

frobware commented:
/retest

frobware commented:
/retest

frobware commented:
/retest

frobware commented:
/retest

frobware commented:
/retest

frobware commented:
/retest

sgreene570 commented on test/extended/router/http2.go:
```diff
@@ -54,29 +61,77 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
+			exutil.DumpPodLogsStartingWithInNamespace("router", "openshift-ingress", oc.AsAdmin())
+		}
+		if len(shardConfigPath) > 0 {
+			oc.AsAdmin().Run("delete").Args("-n", "openshift-ingress-operator", "-f", shardConfigPath).Execute()
```
Should we print an error when deleting here just in case (might be useful when running this test outside of CI).

frobware replied:
Done in 1506505c801a8bd5a37195527b45bf70370a2a93. Thanks.

frobware commented on test/extended/router/http2.go:
```diff
@@ -54,29 +61,77 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
-			e2e.ExpectNoError(e2epod.WaitForPodRunningInNamespaceSlow(oc.KubeClient(), "http2", oc.KubeFramework().Namespace.Name))
+
+			g.By(fmt.Sprintf("creating router shard %q from a config file %q", oc.Namespace(), http2RouterShardConfigPath))
+			shardConfigPath, err = shard.DeployNewRouterShard(oc, 15*time.Minute, shard.Config{
```
Running these previously (when http/2 was enabled by default) I saw lots of flakes <15mins. If anything, I suspect waiting for DNS may need a little longer in CI.

frobware commented on test/extended/router/http2.go:
```diff
@@ -54,29 +61,77 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
+
+			shardedDomain := oc.Namespace() + "." + defaultDomain
+
+			g.By(fmt.Sprintf("creating routes from a config file %q", http2RoutesConfigPath))
```
I don't think it matters. Creating the routes is cheap so whether they go before or after won't really alter how long it takes for a new ingress controller to roll out.

frobware commented:
/retest

frobware commented:
gcp-upgrade: fetch image failed

/retest

frobware commented:
/retest

frobware commented:
/retest

sgreene570 commented on test/extended/testdata/router/router-shard.yaml:
```diff
@@ -0,0 +1,27 @@
+    replicas: 1
+    domain: ${DOMAIN}
+    endpointPublishingStrategy:
+      type: LoadBalancerService
```
This test should probably only ever run on platforms that support LoadBalancerService, right? We may need to include skips for all other platforms at the top of the http/2 test.

frobware replied:
Done in a46c9a5

frobware commented:
/retest

frobware commented:
/test e2e-metal-ipi-ovn-ipv6

frobware commented:
/retest

Miciah commented:
/approve

Miciah commented:
/assign @knobunc

candita commented on test/extended/router/http2.go:
```diff
@@ -54,29 +62,86 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
+			switch infra.Status.PlatformStatus.Type {
+			case configv1.OvirtPlatformType, configv1.KubevirtPlatformType, configv1.LibvirtPlatformType, configv1.VSpherePlatformType:
+				g.Skip("Skip on platforms where the default router is not exposed by a load balancer service.")
+			}
```
I think you could also skip on None and metal.

frobware replied:
Done in https://github.com/openshift/origin/pull/25897/commits/b84944f1116f690213418d6893be1c7353749fa4

candita commented on test/extended/router/http2.go:
```diff
@@ -54,29 +62,86 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
+
+			// certificate start and end time are very
+			// lenient to avoid any clock drift between
+			// between the test machine and the cluster
```
nit - there are 2 "between" s

frobware replied:
Done in https://github.com/openshift/origin/pull/25897/commits/b84944f1116f690213418d6893be1c7353749fa4

candita commented on test/extended/router/http2.go:
```diff
@@ -143,39 +208,58 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
-					var err error
-					url := "https://" + hostname
-					resp, err = client.Get(url)
+				o.Expect(wait.Poll(3*time.Second, 15*time.Minute, func() (bool, error) {
```
Do you think 5 minutes should be enough to wait for a failure? The tests take a pretty long time already to deliver an outcome.

frobware replied:
I think it is something we could monitor and adjust as we see this go through CI on a daily basis. I would rather not make this flake a lot to begin with.

frobware commented:
level=error msg=Cluster initialization failed because one or more operators are not functioning properly.
level=error msg=The cluster should be accessible for troubleshooting as detailed in the documentation linked below,
level=error msg=https://docs.openshift.com/container-platform/latest/support/troubleshooting/troubleshooting-installations.html
level=error msg=The 'wait-for install-complete' subcommand can then be used to continue the installation
level=fatal msg=failed to initialize the cluster: Working towards 4.8.0-0.ci.test-2021-03-25-174822-ci-op-8d0nh089: 687 of 689 done (99% complete), waiting on cloud-controller-manager

Not sure if this is a flake at the moment, but:

/test e2e-aws-fips

candita commented on test/extended/router/shard/shard.go:
```diff
@@ -0,0 +1,90 @@
+		"NAME="+cfg.Name,
+		"NAMESPACE=openshift-ingress-operator",
+		"DOMAIN="+cfg.Domain,
+		"TYPE="+cfg.Type).OutputToFile("config.json")
```
Does it need the OutputToFile?

candita commented:
Just a few nits, otherwise this looks good to me.  
/lgtm

frobware commented:
/retest

frobware commented:
/retest

frobware commented:
/skip

frobware commented:
/retest

frobware commented:
/skip

frobware commented:
/retest

frobware commented:
/retest

frobware commented:
/retest

frobware commented:
/hold 

Need to find out what image I can use when building the server code in the test pod.

frobware commented:
This is ready for review now. 

Leaving the hold as it requires https://github.com/openshift/cluster-ingress-operator/pull/585 and I have pushed a change here that references a pre-built image of 585.

frobware commented:
/skip

frobware commented:
```
Cluster accessed images that were not mirrored to the testing repository or already part of the cluster, see test/extended/util/image/README.md in the openshift/origin repo:

quay.io/frobware/cio from pods:
  ns/e2e-test-router-http2-hlj7h pod/http2 node/ci-op-b9slr0hd-2a78c-jdbhj-worker-c-pktkm

```

The house always wins...

frobware commented:
/hold cancel

Still waiting for openshift/cluster-ingress-operator#585 and I can't test^W cheat my way through that - https://github.com/openshift/origin/pull/25897#issuecomment-811995717

frobware commented:
/retest

frobware commented:
/hold cancel

frobware commented:
/skip

frobware commented:
/retest

candita commented on test/extended/router/certgen/certgen.go:
```diff
@@ -0,0 +1,119 @@
+		IsCA:                  true,
+	}
+
+	rootDerBytes, err := x509.CreateCertificate(rand.Reader, &rootTemplate, &rootTemplate, &rootKey.PublicKey, rootKey)
```
Is it ok for this to be self-signed or do we want a more conventional certificate for testing?

frobware replied:
For testing purposes self-signed is enough; the nature of the test wouldn't materially change and self-signed it easier.

candita commented on test/extended/router/certgen/certgen.go:
```diff
@@ -0,0 +1,119 @@
+// representation (SEC 1, ASN.1 DER form) suitable for dropping into a
+// route's TLS key stanza.
+func MarshalPrivateKeyToDERFormat(key *ecdsa.PrivateKey) (string, error) {
+	data, err := x509.MarshalECPrivateKey(key)
```
Is there any specific reason to use EC rather than RSA keys?

frobware replied:
No, just code that I used first time around when I created the tests.

candita commented on test/extended/router/certgen/certgen.go:
```diff
@@ -0,0 +1,119 @@
+}
+
+// GenerateKeyPair creates root CA, certificate and key with optional
+// hosts. Certificate is valid from notBefore and expires after
```
Are hosts really optional, or does there have to be at least one?

sgreene570 commented on test/extended/router/grpc-interop.go:
```diff
@@ -3,177 +3,202 @@ package router
+				g.Skip("Skip on platforms where the default router is not exposed by a load balancer service.")
+			}
+
+			defaultDomain, err := getDefaultIngressClusterDomainName(oc, 5*time.Minute)
```
5 minutes (or 10 minutes in the other calling site) to get the cluster ingress domain seems very long. How about we set the timeout in both cases to say 1 or maybe 2 minutes?

candita commented on test/extended/router/http2.go:
```diff
@@ -172,25 +255,123 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
+		return "", err
+	}
+	for _, v := range o.Status.Versions {
+		if v.Name == "canary-server" {
```
There is a `const` you can use, `CanaryImageVersionName` in pkg/operator/controller/status/controller.go `status` package.

sgreene570 replied:
Probably not worth vendoring the ingress-operator for this.

frobware replied:
I didn't want to vendor for a single string constant (given that it is unlikely to change).

sgreene570 commented on test/extended/router/http2.go:
```diff
@@ -143,23 +217,32 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
+				o.Expect(err).NotTo(o.HaveOccurred())
+				o.Expect(addrs).NotTo(o.BeEmpty())
+			} else {
+				addrs, err := resolveHost(oc, time.Minute, 15*time.Minute, shardService.Status.LoadBalancer.Ingress[0].Hostname)
```
Same Q as earlier..... does attempting to resolve the LB's hostname tell us anything about DNS for the route?

frobware replied:
My assumption and, in practice, assertion is that as soon as <anything>.new-domain resolves so will route name.

frobware replied:
As we discussed offline will also validate the DNS of a route for the case where we have a hostname after the LB name resolves.

candita commented on test/extended/router/grpc-interop.go:
```diff
@@ -3,177 +3,202 @@ package router
-			Insecure: true,
-		}
+		if err := wait.PollImmediate(time.Second, timeout, func() (bool, error) {
+			e2e.Logf("Dialling: %+v", dialParams)
```
nit - US English spells this as "Dialing"

candita commented on test/extended/router/grpc-interop.go:
```diff
@@ -3,177 +3,202 @@ package router
-	}
-}
+	for _, name := range testCases {
+		e2e.Logf("Running gRPC interop test case %q via host %q", testCases, dialParams.Host)
```
`testCases` should be `name` here

candita commented on test/extended/router/grpc-interop.go:
```diff
@@ -3,177 +3,202 @@ package router
-				err := grpcinterop.ExecTestCase(conn, name)
-				o.Expect(err).NotTo(o.HaveOccurred())
+			if err := grpcinterop.ExecTestCase(conn, name); err != nil {
+				e2e.Logf("error: running gRPC interop test case %q through %q: %v, retrying...", testCases, dialParams.Host, err)
```
`testCases` should be `name` here

frobware commented on test/extended/router/grpc-interop.go:
```diff
@@ -3,177 +3,202 @@ package router
+					o.Expect(err).NotTo(o.HaveOccurred())
+					o.Expect(addrs).NotTo(o.BeEmpty())
+				} else {
+					addrs, err := resolveHost(oc, time.Minute, 15*time.Minute, shardService.Status.LoadBalancer.Ingress[0].Hostname)
```
We only want to continue hitting the route when the route actually resolved to the new LB's IP address. We loop until that condition is met. We could proceed as the test itself also loops but you get a lot of noise logged because it is hitting the default ingress controller and that doesn't have http/2 enabled. As soon as <any-name>.new-domain resolves we know the route hostname will too. Too subtle?

candita commented on test/extended/router/http2.go:
```diff
@@ -172,25 +255,123 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
 				o.Expect(resp.Proto).To(o.Equal(tc.frontendProto), testConfig)
 				body, err := ioutil.ReadAll(resp.Body)
 				o.Expect(err).NotTo(o.HaveOccurred(), testConfig)
+				resp.Body.Close()
```
Should this `Close()` be done as a defer to make sure it always happens?

frobware replied:
I've removed this as we assert with a close a few lines below anyway.

candita commented on test/extended/router/http2.go:
```diff
@@ -172,25 +255,123 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
 				o.Expect(err).NotTo(o.HaveOccurred(), testConfig)
+				resp.Body.Close()
 				o.Expect(string(body)).To(o.Equal(tc.backendProto), testConfig)
 				o.Expect(resp.Body.Close()).NotTo(o.HaveOccurred())
```
This seems odd to check, but if you need to, then disregard my other comments about `resp.Body.Close`

frobware replied:
Yeah, there should only have been one call - and this one was intentional.

candita commented on test/extended/router/http2.go:
```diff
@@ -172,25 +255,123 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
+		}
+
+		for i := range addrs {
+			if addrs[i] == expectedAddr {
```
Why index `addrs` instead of using direct assignment in the `for` statement?

frobware replied:
Just cheaper.

candita commented on test/extended/testdata/router/router-grpc-interop-routes.yaml:
```diff
@@ -0,0 +1,48 @@
+    port:
+      targetPort: 8443
+    tls:
+      termination: passthrough
```
No edge terminated route needed in this config?

frobware replied:
I simplified these tests to as little as possible taking place between the client call and backend. If we think it is worthwhile we add edge and reencrypt later.

frobware replied:
> No edge terminated route needed in this config?

Edge isn't supported. 

 This still exists but I removed the route: 

```go
				if routeType == routev1.TLSTerminationEdge {
					e2e.Logf("skipping %v tests - needs https://github.com/openshift/router/pull/104", routeType)
					continue
				}

```

I can put the edge route back.

frobware replied:
I put the edge route back.

candita commented on test/extended/testdata/router/router-shard.yaml:
```diff
@@ -0,0 +1,27 @@
+apiVersion: template.openshift.io/v1
+kind: Template
+parameters:
+- name: NAME
```
The NAME parameter is not used in here.

candita commented on test/extended/testdata/router/router-grpc-interop.yaml:
```diff
@@ -129,190 +25,30 @@ objects:
-          port: 8080
-        initialDelaySeconds: 10
-        periodSeconds: 3
+      command: ["ingress-operator", "serve-grpc-test-server"]
```
There is another subcommand `serve-http2-test-server` in router-http2.yaml -- are there 2 new subcommands?

candita replied:
Yes, I see there are.

candita replied:
These are already in cmd aren't they?

frobware replied:
There are two new commands, one serves the tests in http2.go and the the is for the grpc-interop tests.

candita commented on test/extended/testdata/router/router-http2.yaml:
```diff
@@ -45,168 +29,35 @@ objects:
+    - image: ${IMAGE}
       name: server
-      command: ["/workdir/http2-server"]
+      command: ["ingress-operator", "serve-http2-test-server"]
```
There is another subcommand `serve-grpc-test-server` in router-grpc-interop.yaml

candita replied:
Working as defined :)

frobware commented on test/extended/router/http2.go:
```diff
@@ -172,25 +255,123 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
 						return false, nil // could be 503 if service not ready
 					}
 					if tc.statusCode == 0 {
+						resp.Body.Close()
```
On the happy path I want to assert that calling close does not error. Using defer would prevent that.

frobware commented:
@sgreene570 @candita I addressed your review feedback in commits:

- https://github.com/openshift/origin/pull/25897/commits/881e34273f0b40f12d889f00b71d55670066fa8d
- https://github.com/openshift/origin/pull/25897/commits/dcf76b1c7a2bfd3fe3d199ae65bffcbde282122c
- https://github.com/openshift/origin/pull/25897/commits/56b4827b48503188258361149a67fd13016dc611
- https://github.com/openshift/origin/pull/25897/commits/57d26b506d4f0803568577539187c4950df488e3

sgreene570 commented on test/extended/router/grpc-interop.go:
```diff
@@ -3,177 +3,206 @@ package router
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Getting LB service")
+			shardService, err := getRouterService(oc, 10*time.Minute, "router-"+oc.Namespace())
```
10 minutes to get the ingress controller's service seems high.

frobware replied:
Dropped to 5 mins in https://github.com/openshift/origin/pull/25897/commits/1f2980383228d53218a0f8ba266ed89d7b976734.

@sgreene570 if you have no other requests I propose to squash and we can move this forward.

sgreene570 replied:
Ack. Squash away!

frobware replied:
Done.

frobware commented:
/skip

sgreene570 commented:
/lgtm

frobware commented:
[sig-scheduling] Multi-AZ Clusters should spread the pods of a service across zones [Serial] [Suite:openshift/conformance/serial] [Suite:k8s] expand_less | 11m24s

fail [k8s.io/kubernetes@v1.20.0/test/e2e/scheduling/ubernetes_lite.go:65]: Unexpected error:     <*errors.errorString \| 0xc00264e0a0>: {         s: "Error waiting for 1 pods to be running - probably a timeout: Timeout while waiting for pods with labels \"startPodsID=717a9466-7a6e-4433-8419-c4a42d8a0d47\" to be running",     }     Error waiting for 1 pods to be running - probably a timeout: Timeout while waiting for pods with labels "startPodsID=717a9466-7a6e-4433-8419-c4a42d8a0d47" to be running occurred

/retest

frobware commented:
/skip

knobunc commented:
/approve

frobware commented:
/skip

frobware commented:
/retest

frobware commented:
gcp-upgrade may be this: https://bugzilla.redhat.com/show_bug.cgi?id=1926269

/retest

frobware commented:
/retest

frobware commented:
/retest

frobware commented:
/retest

frobware commented:
failed to acquire lease: resources not found
Not sure that's for all the failures, ...

/retest

frobware commented:
/retest

frobware commented:
/retest

Miciah commented:
[sig-instrumentation] Prometheus when installed on the cluster should start and expose a secured proxy and unsecured metrics [Skipped:Disconnected] [Suite:openshift/conformance/parallel]
    fail [github.com/openshift/origin/test/extended/prometheus/prometheus.go:344]: some services expose metrics over insecure channel
    Expected
        <[]error | len:6, cap:8>: [
            {
                s: "following target does not secure metrics endpoint: openshift-monitoring/promtail-monitor",

/test e2e-gcp

frobware commented:
[sig-instrumentation] Prometheus when installed on the cluster should start and expose a secured proxy and unsecured metrics [Skipped:Disconnected] [Suite:openshift/conformance/parallel] expand_more | [sig-instrumentation] Prometheus when installed on the cluster should start and expose a secured proxy and unsecured metrics [Skipped:Disconnected] [Suite:openshift/conformance/parallel] expand_more


[sig-instrumentation] Prometheus when installed on the cluster should start and expose a secured proxy and unsecured metrics [Skipped:Disconnected] [Suite:openshift/conformance/parallel] expand_more
: [sig-instrumentation][Late] Alerts shouldn't report any alerts in firing or pending state apart from Watchdog and AlertmanagerReceiversNotConfigured and have no gaps in Watchdog firing 

[Suite:openshift/conformance/parallel] expand_more | : [sig-instrumentation][Late] Alerts shouldn't report any alerts in firing or pending state apart from Watchdog and AlertmanagerReceiversNotConfigured and have no gaps in Watchdog firing [Suite:openshift/conformance/parallel] expand_more

[sig-instrumentation][Late] Alerts shouldn't report any alerts in firing or pending state apart from Watchdog and AlertmanagerReceiversNotConfigured and have no gaps in Watchdog firing [Suite:openshift/conformance/parallel] expand_more

[sig-network] pods should successfully create sandboxes by not timing out | : [sig-network] pods should successfully create sandboxes by not timing out

[sig-network] pods should successfully create sandboxes by not timing out

/retest

frobware commented:
/retest
